suite-common/thp/src/thpReducer.ts

```diff
@@ -0,0 +1,109 @@
+export type ThpState = {
+    step: ThpStep;
+    lastThpCode?: string;
+    credentials: ThpSuiteCredentials[];
+};
+
+const initialState: ThpState = {
+    step: null,
+    lastThpCode: undefined,
+    credentials: [] as ThpSuiteCredentials[],
+};
+
+export const prepareThpReducer = createReducerWithExtraDeps<ThpState>(
+    initialState,
+    (builder, extra) =>
+        builder
+            .addCase(thpActions.resetThpFlow, state => {
+                state.step = null;
+            })
+            .addCase(thpActions.invalidCode, state => {
+                state.step = 'CodeInvalid';
+            })
+            .addCase(thpActions.setLastThpCode, (state, { payload }) => {
+                state.lastThpCode = payload.code;
+            })
+            .addCase(thpActions.showAutoconnectInfo, (state, { payload }) => {
+                const credentialToUpdate = state.credentials.find(
+                    it => it.credential == payload.credential.credential,
+                );
+
+                if (credentialToUpdate !== undefined) {
+                    credentialToUpdate.wasUserAskedToAutoconnect = true;
+                }
+
+                state.step = 'AutoconnectInfo';
+            })
+            .addCase(thpActions.incrementCredentialConnectionCounter, (state, { payload }) => {
+                const credentialToUpdate = state.credentials.find(
+                    it => it.credential == payload.credential.credential,
+                );
+
+                if (credentialToUpdate !== undefined) {
+                    credentialToUpdate.connectionCounter = credentialToUpdate.connectionCounter + 1;
+                }
+            })
+            .addMatcher(
+                action => action.type === UI.REQUEST_THP_PAIRING,
+                state => {
+                    state.step = 'CodeEntry';
+                },
+            )
+            .addMatcher(
+                action => action.type === DEVICE.THP_CREDENTIALS_CHANGED,
+                (state, action: DeviceThpCredentialsChanged) => {
+                    const { credentials } = action.payload;
+
+                    state.credentials.push({
+                        ...credentials,
+                        connectionCounter: 0,
+                        wasUserAskedToAutoconnect: false,
+                    });
+                },
+            )
+            .addMatcher(
+                action => action.type === UI.REQUEST_BUTTON,
+                (state, action: AnyAction) => {
+                    switch (action.payload.name) {
+                        case 'thp_pairing_request':
+                            state.step = 'Pairing';
+                            break;
+                        case 'thp_autoconnect_credential_request':
+                            state.step = 'Autoconnect';
+                            break;
+                        case 'thp_connection_request':
+                            state.step = 'Connection';
+                            break;
```

Comment:
is there a better solution then `AnyAction`? Because that has all properties `any`.
If not, than at least exhaustive typing for the switch (plus comment above):
```suggestion
                    const actionName: THPButtonRequestName = action.payload.name;
                    switch (actionName) {
                        case 'thp_pairing_request':
                            state.step = 'Pairing';
                            break;
                        case 'thp_autoconnect_credential_request':
                            state.step = 'Autoconnect';
                            break;
                        case 'thp_connection_request':
                            state.step = 'Connection';
                            break;
                        default:
                            exhaustive(actionName);
```

Reply:
fixed